File: clojure/day14.clj
(ns day14
  (:require aoc
            [clojure.data.int-map :refer [dense-int-set]]))


(def ^:const start-x 500)
(def ^:const X 256)


(defn down  ^long [^long pt] (inc pt))
(defn left  ^long [^long pt] (- (inc pt) X))
(defn right ^long [^long pt] (+ (inc pt) X))

(defn drop-grain ^long [settled]
  (loop [pt (* X start-x)]
    (let [d (down pt)
          l (left pt)
          r (right pt)]
      (cond
        (not (settled d)) (recur d)
        (not (settled l)) (recur l)
        (not (settled r)) (recur r)
        :else pt))))

(defn fill-to-full [settled pt]
  (if (settled pt) settled
      (-> settled
          (fill-to-full (down pt))
          (fill-to-full (left pt))
          (fill-to-full (right pt))
          (conj pt))))

(defn part-1 [rocks]
  (let [^long floor (reduce max (map #(mod % X) rocks))]
    (loop [settled rocks]
      (let [pt (drop-grain settled)]
        (if (= (mod pt X) (dec floor))
          (- (count settled) (count rocks))
          (recur (conj settled pt)))))))

(defn part-2 [rocks]
  (-> rocks
      (fill-to-full (* X start-x))
      count
      (- (count rocks))))


(defn path->rock [[ax ay] [bx by]]
  (for [x (range (min ax bx) (inc (max ax bx)))
        y (range (min ay by) (inc (max ay by)))]
    (+ (* X x) y)))

(defn add-floor [rocks]
  (let [floor-y (+ 2 (reduce max (map #(mod % X) rocks)))
        floor (for [x (range (- start-x floor-y) (+ start-x floor-y 1))] (+ (* X x) floor-y ))]
    (into rocks floor)))

(defn parse-input [input]
  (->> input
       aoc/read-input
       (into (dense-int-set)
             (comp
              (map aoc/integers)
              (map #(partition 2 %))      ; points
              (mapcat #(partition 2 1 %)) ; paths
              (mapcat #(apply path->rock %))))
       add-floor))

(defn solve
  ([] (solve 14))
  ([input]
   (let [rocks (parse-input input)]
     [(part-1 rocks)
      (part-2 rocks)])))


(solve)

File: python/day14.py
from aoc import *
from itertools import count


def parse_line(line):
    points = mapl(integers, line.split(" -> "))
    rock = set()
    for (ax, ay), (bx, by) in zip(points, points[1:]):
        for px in range(min(ax, bx), max(ax, bx)+1):
            for py in range(min(ay, by), max(ay, by)+1):
                rock.add(complex(px, py))
    return rock


def parse_input(data):
    rock = set()
    for line in data:
        rock |= parse_line(line)
    return rock


START = 500


def solve(file=14):
    def pour(pt, p1):
        if pt in settled:
            return pt.imag == floor
        for dx in (0, -1, 1):
            if pour(pt + dx+1j, p1) and p1:
                return True
        settled.add(pt)
    count_sand = lambda settled: len(settled) - rock_count

    settled = parse_input(read_input(file))
    floor = int(max(pt.imag for pt in settled)) + 2
    for x in range(START-floor, START+floor+1):
        settled.add(complex(x, floor))
    rock_count = len(settled)
    pour(START, True)
    p1 = count_sand(settled)
    pour(START, False)
    return p1, count_sand(settled)


print(solve())










def test():
    assert solve("14_test") == (24, 93)
    assert solve() == (1003, 25771)
